### PR TITLE
msm5232: Envelope generator resistance values can be specified externally.

### PR DESCRIPTION
**Pull request by yaei**
msm5232:
Envelope generator resistance values can be specified externally.
Separated the resistance value for decay for short period and long period.

The sound chip msm5232 calculates how attack and decay proceed in envelope generation using the capacitance of the externally connected capacitor and some internal resistance values.

In the current implementation, the msm5232 internal resistances for envelope generation is fixed values embedded in the source code.
This commit makes these values externally configurable.
The msm5232 operates exactly the same as before, unless the resistance value is specified externally.

Additionally,
msm5232's envelope generation uses two resistance values to determine a decay period.
The current implementation simplifies this process using a single resistance value.
This commit separates the resistance value that msm5232 uses to determine a decay period.
Again, msm5232 emulation works exactly the same as before, unless resistance values are specified externally.

The 3 types of resistance values for envelope generation that msm5232 has inside (1 for attack, 2 for decay) can be calculated from the capacitance of the connected capacitors and the decay period for each parameter value specified described in the document.
So, I made a function that takes the capacitance of the capacitor connected to msm5232 as an argument and sets the resistance values.

This commit does not actually call the functions added in this commit.
Therefore, this commit doesn't change MAME's output in any way.

### CODE REVIEW

**Comment by yaei** [comment]
The reason for this commit is that the sound of the driver flstory (Taito's The Fairyland Story) decays much longer than the real PCB.
This problem was reported 11 years ago by Fujix on MameTesters. https://mametesters.org/view.php?id=4567
In March of this year, ToremiRodomi gave MameTesters a close-up of a PCB capacitor. According to it, the capacitance of the capacitor on the PCB is 1uF.
In flstory's implementation, the capacitance value is correctly set to 1uF.
The decay period of the sound generated by the msm5232 is determined by the capacitance of the capacitor connected to the msm5232 and the resistance values of the internal resistor of the msm5232.
Since the capacitance of the capacitor was confirmed to be the correct value, the long flstory decay is due to the internal resistance being set higher than it actually is in the msm5232 emulation.

Once this commit is approved, I will commit the implementation of the driver(flstory) setting the proper resistance values for msm5232 and, a more minor fix to the issue affecting the reproduction of the decay duration.

**Comment by yaei** [comment]
[5232.pdf](https://github.com/mamedev/mame/files/11213677/5232.pdf)
.pdf attached are the referenced document.
I obtained it from the web and it is not corroborated.
In page 18 (p.17), after ( 2 ) Envelope generator, it describes how attack and decay are obtained from internal resistances.
The figures on page 21 (p.20) and 25 (p.24) are also relevant.

File: src/devices/sound/msm5232.cpp [hunk 1]
```diff
@@ -5,6 +5,9 @@
 
 #define CLOCK_RATE_DIVIDER 16
 
+#define R51 1400    /* charge resistance */
```
Comment (rb6502) [inline src/devices/sound/msm5232.cpp:8]: These constants should be turned into **constexpr int** instead of #defines.

File: src/devices/sound/msm5232.cpp [hunk 2]
```diff
@@ -209,8 +214,8 @@ static FILE *sample[9];
  */
 
 
-#define R51 1400    /* charge resistance */
-#define R52 28750   /* discharge resistance */
+//#define R51 1400    /* charge resistance */
```
Comment (rb6502) [inline src/devices/sound/msm5232.cpp:217]: Please delete lines that are no longer used.

**Comment by yaei** [comment]
Thank you for the review.
I have fixed the problems.

**Comment by yaei** [comment]
MSM5232 generates attack and decay waveforms from the voltage transitions associated with charging and discharging a connected capacitor.
The voltage of the connected capacitor is treated as the volume as it is.
In general, the voltage transition of a capacitor is determined by the capacitance of the capacitor and the resistance of the connection.
In the MSM5232, the resistance of the connection is a fixed value built into the IC.

As you said, these three resistance values are built into the MSM5232 as fixed values and cannot be changed externally.
However, under the current implementation of MAME, it is not possible to uniquely determine what these values are.
I will attempt to explain why this commit took this form.
Basically, the purpose of this commit is, to minimize the scope of influence while making behavior of a particular driver closer to the actual machine.

**Comment by yaei** [comment]
In some schematics, use cases, and the actual Korg Poly-800 instance I have, there is no external resistor associated with the capacitor connected to the msm5232.
MSM5232 works without any external resistors connected to those directly connected capacitors.

Some examples:
Schematic of a TAITO title using MSM5232.
Buggy Challenge (buggychl)
https://www.arcade-museum.com/manuals-videogames/B/BuggyChallenge.sch.pdf
The sound board starts on p. 19.
Unfortunately, the circuitry around the MSM5232 is not included.
The capacitance of the connected capacitors is 1.0uF.

Manual for the KORG PSS-50, with a circuit diagram using MSM5232.
https://archive.org/details/KORGPSS50/page/n4/mode/1up
The capacitance of the connected capacitors is 1.0uF.
1u/50V RB LL x 8

MSM5232 driven by Arduino. Schematic is available.
https://oykenkyu.blogspot.com/2022/03/MSM5232RS.html

**Comment by yaei** [comment]
There is some confusion in the driver that uses msm5232 included in mame.
The following is taken from a slightly older MAME (0.223?) source codes.
Looking at the driver implementations, there are three types of capacitance values specified:
	0.39e-6, 0.65e-6, 1.0e-6
	buggychl.cpp(623):	m_msm->set_capacitors(0.39e-6, 0.39e-6, 0.39e-6, 0.39e-6, 0.39e-6, 0.39e-6, 0.39e-6, 0.39e-6); /* default 0.39 uF capacitors (not verified) */
	wyvernf0.cpp(674):	msm.set_capacitors(0.39e-6, 0.39e-6, 0.39e-6, 0.39e-6, 0.39e-6, 0.39e-6, 0.39e-6, 0.39e-6); /* default 0.39 uF capacitors (not verified) */	
	bigevglf.cpp(428):	m_msm->set_capacitors(0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6f, 0.65e-6, 0.65e-6); /* 0.65 (???) uF capacitors */
	ladyfrog.cpp(328):	m_msm->set_capacitors(0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6);
	msisaac.cpp(488):	m_msm->set_capacitors(0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6); /* 0.65 (???) uF capacitors (match the sample, not verified) */
	nycaptor.cpp(786):	m_msm->set_capacitors(0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6); /* 0.65 (???) uF capacitors (match the sample, not verified) */
	nycaptor.cpp(856):	m_msm->set_capacitors(0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6); /* 0.65 (???) uF capacitors (match the sample, not verified) */
	nycaptor.cpp(923):	m_msm->set_capacitors(0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6, 0.65e-6); /* 0.65 (???) uF capacitors (match the sample, not verified) */
	flstory.cpp(786):	m_msm->set_capacitors(1.0e-6, 1.0e-6, 1.0e-6, 1.0e-6, 1.0e-6, 1.0e-6, 1.0e-6, 1.0e-6); /* 1.0 uF capacitors (verified on real PCB) */

0.39e-6 is, the value listed as the standard capacity in the msm5232 datasheet can be found on the web.
In the current implementation of MAME, if the driver specifies this value as the capacitance, the output decay duration is noticeably shorter for the real PCB.
buggychl, wyvernf0 uses this value. So, in the output from these drivers, the duration of decays are shorter.

0.65e-6 is, a value that, when set for the current MAME implementation, produces an output close to that of the real PCB.
While it is a value that produces an output that feels audibly correct, but the rationale cannot be explained.

1.0e-6 is, the value of the capacitor built into the board of the flstory real PCB.
If this value is specified for the current implementation of MAME, the decay duration is obviously longer for the real PCB.
This is the main reason for the longer decay of output from flstory.

**Comment by yaei** [comment]
The resistance values in question can be calculated from the ideal attack/decay duration and the capacitance of the capacitor connected to the msm5232.

Based on the msm5232 datasheet which can be found on the web (.pdf I have attached above),
the ideal decay duration for Decay time data $0 is 40ms.
The capacitance of the capacitor to be connected is 0.39uF.
So, the resistance value for decay can be calculated to be 44543.023 ohms using the following:
```
Vt = V0 * exp(-t/RC)
R = -t/ ( C*log(Vt/V0) )
-0.04 /(0.39e-6*log( 1-0.9 )) = 44543.0237849
```
If this resistance value is adopted as the embedded value that the msm5232 emulation code has, drivers which use MSM5232 must specify 0.39uF as the capacitance connected to the msm5232 to obtain correct decay duration.
If 0.65uF or 1.0uF is specified, the decay duration will be obviously long relative to the output from the real PCB.
We have confirmed that the capacitor on flstory's actual PCB is 1.0uF, so this resistance value is clearly wrong at least for the msm5232 used by flstory.

**Comment by yaei** [comment]
Some parts are repetitive, but I write a little more:

According to the msm5232 datasheet, the decay period is specified by the lower four bits of addr $0A and $0B. (p12-15)
When these 4 bits are 0***, the first of the resistance values for decay is used to discharge the externally connected capacitor. (p17-18)
Decay duration, the time it takes for the capacitor to discharge to 10%, is specified as 40 ms when these four bits are 0000.
This is consistent with the output of the actual machine. (p20,24)

The ideal resistance value to discharge a 1.0uF capacitor to 10% in 40ms is 17371.779 Ohms.
-0.04 /(1.0e-6*log( 1-0.9 ) ) = 17371.7792761
The fixed value that the msm5232 emulation code currently has is greater than this(28750.0 Ohms).
The rationale for this value (28750 Ohms) is not known.
The comments attached to the msm5232 code suggest that it is an approximation from the capacity described in the data sheet.

The value 0.65uF, which some drivers specifies as capacitance, provides decay durations close to the real PCBs, when specified by a title with a 1.0uF capacitor connected on the actual board.
This is because the current implementation has a resistance value of 28750 ohms for some reason. There is no rationale for this capacitance value.

Following formula explains that specifying 0.65uF for the current msm5232 implementation will result in a decay duration close to that of the actual device.
exp(-0.04/(28750.0*0.65e-6)) = 0.11759976669	// Discharging a fully charged 0.65uF capacitor with a resistance of 28750 Ohms for 0.04 seconds results in a voltage of 11.75%
The ideal value of this formula is 0.1, so to be exact, a driver with a capacitance value of 0.65uF should have a slightly longer decay duration than the actual PCB.

if we follow the datasheet found on the web, the capacitance of the connected capacitor is 0.39uF, so this resistance value must be 44543.023 Ohms.
-0.04 /(0.39e-6*log( 1-0.9 ) ) = 44543.0237849
When using this value, each driver must specify 0.39uF as the capacitance to obtain the correct output.
Specifying 0.65uF or 1.0uF will result in significantly longer decay durations.

**Comment by yaei** [comment]
What can be assumed from the above,
these resistance values may vary with different variations of msm5232, or the datasheet does not match the actual situation.

At least for The Fairyland Story(flstory) and Buggy Challenge(buggychl),
we can confirm that 1.0uF capacitors are connected to the MSM5232, so the resistance values below can be adopted.
    -0.002/(1.0e-6*ln( 1-0.9 ) ) = 868.588963807
    -0.04 /(1.0e-6*ln( 1-0.9 ) ) = 17371.7792761
    -0.25 /(1.0e-6*ln( 1-0.9 ) ) = 108573.620476
These resistance values are those that would produce a plausible output if 1.0uF were specified as the capacitance for all of the drivers cited above.

Personally, I think these values could be embedded in the msm5232 code rather than specified externally, but there is no verification to back this up to do so.
I am unable to check the capacitors on the actual PCB for all other drivers that use msm5232 and compare their outputs with those of mame.
It is still possible that the resistance values vary with msm5232 variants.

That is all for this explanation.

If we implement that the resistance values can be specified for each driver, it may be better to specify them in parameters passed to the constructor of msm5232_device, instead of calling a function afterwards.

**Comment by cuavas** [comment]
So basically, that’s a lot of words to say this is a mechanism for adding per-game hacks.

There’s no way for an external circuit to change the characteristics of the attack, decay and release drivers, as they’re part of the chip.

An series resistor can be added, but that will change the characteristics substantially because the voltage drop across the resistor will change polarity when the envelope switches from attack to decay and the current changes direction.  Adding a series resistor will cause a sudden drop in volume on the transition from attack to decay.  The series resistor will also affect attack, decay and release equally.

You could conceivably add two separate series resistors in series with diodes to alter the attack and decay/release time independently, but then you’d get an even bigger instantaneous change in volume due to the junction bias voltage of the diodes.  I very much doubt anyone would do this.

Anyway, that’s all beside the point, as if I understand correctly, the game in question just has capacitors connected to the envelope generator pins.

One thing you don’t seem to have considered at all is component aging effects.  IIRC the game in question uses dry sintered tantalum capacitors.  Cheaper tantalum capacitors can develop increased leakage current as they age (or develop short circuits in extreme cases), which would affect the time constant.  How old are the capacitors on the boards you recorded audio from?

At any rate, this PR is not a suitable approach.  Either the chip is not characterised properly, or there are are other factors affecting the behaviour.  Facilitating per-game hacks holds back emulation.

**Comment by yaei** [comment]
Thanks for reading.
I am aware of the deterioration of capacitors.
However, the output produced by MAME in its current state is longer than can be explained by capacitor degradation in the actual device.
I have heard the sound of the actual machine in the eighties when the title was released, and have heard the official soundtrack etc.

I understand that putting in some hack for each driver is evil idea.
I also agree with not allowing external parameters to be specified that are not in the original chip.

The fixed values that the current msm5232 emulation has are wrong I believe, but I don't have the backing to correct them.
In this case, what kind of fix would be adoptable?
One way to do this,
I am wondering if I can run the msm5232 on a real machine and prove the resistance values to rewrite the fixed values that the emulation of the msm5232 currently has.
Or explain that the capacitors are degraded and set the capacitance value set in flstory to 0.65e-6. By doing so, we can obtain a reproducibility close to the actual PCB. However, I believe this is not what is actually happening.

Anyway, thank you for taking the time to respond.

**Comment by cuavas** [comment]
Well, we need someone to characterise the chip in the three envelope phases (attack, decay and release).  Assuming it behaved linearly, you could measure the open-circuit voltage on one of the the pins in each phase to get the effective drive voltage, and then wire a voltage divider to the pin so you can measure the current into a known impedance (it can be inferred by measuring the voltage).

**Comment by TremiRodomi** [comment]
Hi, I'm the person who provided additional verification for Fairyland Story's PCB values, and a self-proclaimed expert on the game in general. 

The aging components issue, whole a valid point, doesn't hold much value here. Even moreso when there's an overwhelming amount of evidence that MAME's sound emulation of the game, with evidence going back from near the game's release, is incorrect in general. An official recording of the game's music done for Taito Game Music (attached below), an album made in 1987, has note decay similar to verified audio/footage of the PCB, stereo sweetening and lower pitch than usual aside. 

While I also agree that per-game hacks typically isn't the right approach to accurate emulation, for now I think the proposed solution it would be a nice stopgap for now until someone completely verifies the internal impedances of the chip since it only affects games on this driver. Nevermind the fact that MAME already has a lot of per-game hacks on a lot of games...

https://youtu.be/IFhJv-Zz6nM?t=891

**Comment by cuavas** [comment]
The trouble is, you accept one of these as a "stopgap" and it opens the floodgates for a pile of contradictory per-game hacks based on what "sounds right" to different people. It's exactly the kind of thing that we'd be stuck with forever. There are too many of these in MAME already.

If the envelope generator isn't characterised properly in MAME, it needs to be characterised properly - that requires about eight measurements.

**Comment by yaei** [comment]
Thank you both for getting involved. I'm glad you did.

I understand that the fatal flaw in this commit was that it added an implementation to the chip code that the actual chip did not have in order to fix one particular driver.
I'll make another, much simpler pull request about flstory's driver in the proper manner.
I don't expect that to be easily accepted, but I will try.

In addition to that,
I already have a few variants of MSM5232 on hand, so I will try to get them to work.
And, I'll ask OKI Electric about it. It should be better than not doing it.

**Comment by TremiRodomi** [comment]
Greetings once again. I believe cuavas misread me. While I do agree that this is not the best approach in accurately fixing the issue for the moment, I think it is a good temporary fix until someone properly finds the actual values in the MSM5232 chip, since the fix only applies to this driver. On that note, eito/yaei said that there are several variants of the chip, implying that this fix might actually make the most sense anyhow. This isn't making it ""sounds right" to different people" (sic), this is objectively what the game is supposed to sound like. Most of the official soundtracks insinuate it as such, footage/audio from a number of PCBs insinuate it as such, and even the contemporary home ports of the game point towards the direction of the low note decay. Even the Arcade Archives and the Egret II Mini versions of the game, while otherwise inaccurate, also point towards that direction. For further evidence, I will even happily record footage from my own PCB of the game once I get my supergun set up properly, if requested.

In addition, MAME already has a bunch of per-game stopgaps for the sake of accuracy, so I think it is rather baffling to say that MAME doesn't take those in. A number of games on Space Invaders-like hardware still use the Space Invaders samples since schematics for their audio circuits haven't been found yet to fully confirm what they actually sound like. Hell, a lot of the Taito F3 driver was mainly built on guesses until somewhat recently, and even then a lot of the driver still relies on said guesses. Etc.

For such an obscure game like this, I do not mind a guess or two in order to bring it more with PCB parity, especially since detailed schematics of the game and the MSM5232 chip aren't freely available. If there was I'd surely point towards the information, but alas. Sure, it's not the best solution, but it's nice for now and will still bring the game closer to how it's supposed to be compared to the current implementation.

**Comment by cuavas** [comment]
No, it doesn't only apply to one driver. It adds an interface vote per-game hacks to the sound chip. That is not a good idea, and every time something like this is added as a "temporary solution", we end up stuck with it for far too long. In this case, it would encourage people to adjust the numbers to make games "sound right" rather than actually characterising the chip properly.

**Comment by galibert** [comment]
Configurable resistances make no sense, because they're on-die.

Now, the datasheet specifies respectively 2ms, 40ms and 0.33ms with a 0.39uF capacitor, for resistances of 2230, 44500 and 367 ohms respectively.  For some mysterious reasons the driver uses 1400 and 28750 (and forgets about the 3rd?) which would reach the datasheet timings with a capacitance of 0.62uF, which is damn close to 0.65uF.

But that doesn't mean the datasheet is right.  In particular all the examples we seem to find in the real world uses 1uF capacitors.  That makes me think that oki retuned the chip for that value and told its clients without changing the spec.  So personally I'd use the datasheet timings with a 1uF cap instead of 0.39uF, e.g. resistances of 870, 17400, and 140 ohms.  It's easier to do lower resistances on a die anyway, so that could be a reason.

And then I'd change all the games to 1uF and see how it goes.  But per-game configurability of the resistances makes no sense unless we find a pcb with resistances added in series.

**Comment by yaei** [comment]
Or even simpler, just setting the capacitance value in flstory to 0.65e-6 would satisfy the original motivation.
I don't think that is a good way.
I would still be happy if it passes, though.

**Comment by galibert** [comment]
We know for a fact that putting flstory at 0.65 is wrong, so we try to avoid that.

**Comment by TremiRodomi** [comment]
That's more like it. I've always noticed the MSM5232 emulation in MAME was a bit off in general and this might be why. It's with Fairyland Story I noticed it even more. 

Glad to see an actual solution is being worked out now!

**Comment by yaei** [comment]
I didn't expect it, fascinating, thank you!
Sure, I will create a PR dealing with the 0x6/0x7 issue.

**Comment by galibert** [comment]
Oh, will you want yaei or eito or something else for the credit?  I can still change it.

**Comment by yaei** [comment]
I don't need it.
Thank you for confirming!

**Comment by TremiRodomi** [comment]
Wonderful.

**Comment by galibert** [comment]
Obsoleted by other PRs
